[SobolEngine] Update direction numbers to 21201 dims

Balandat opened the pull request:
Performs the update that was suggested in https://github.com/pytorch/pytorch/issues/41489

Adjust the functionality to largely match that pf the scipy companion PR https://github.com/scipy/scipy/pull/10844/, including
- a new `draw_base2` method
- include zero as the first point in the (unscrambled) Sobol sequence

The scipy PR is also quite opinionated if the `draw` method doesn't get called with a base 2 number (for which the resulting sequence has nice properties, see the scipy PR for a comprehensive discussion of this). 

Note that this update is a **breaking change** in the sense that sequences generated with the same parameters after as before will not be identical! They will have the same (better, arguably) distributional properties, but calling the engine with the same seed will result in different numbers in the sequence.

Test Plan:
```
from torch.quasirandom import SobolEngine

sobol = SobolEngine(3)
sobol.draw(4)

sobol = SobolEngine(4, scramble=True)
sobol.draw(5)

sobol = SobolEngine(4, scramble=True)
sobol.draw_base2(2)
```

Differential Revision: D25657233

malfet reviewed (changes requested):
All Sobol related tests are failing, can you please look at those (and either update the tests or fix underlying failures)

Balandat commented:
I didn't touch the tests yet b/c I wanted to get a read on how reasonable putting the constants into the humongous constant C-array. If that seems reasonable I'm happy to fix up the units.

Balandat commented:
@malfet I updated the PR so that it implements the same behavior as the Sobol sampler in the scipy PR - the implementation and statistical properties were discussed at length with experts in the field there (443 comments on the PR and counting), so I am very confident about this being the correct thing to do. 

I adapted the tests from the scipy PR. I can add more of the old tests back in, but I will have to update target numbers b/c of the breaking change to the direction numbers and the inclusion of zero as the first point.

malfet commented on aten/src/ATen/native/SobolEngineOps.cpp:
```diff
@@ -126,19 +126,28 @@ Tensor& _sobol_engine_initialize_state_(Tensor& sobolstate, int64_t dimension) {
-  for (int64_t d = 0; d < dimension; ++d) {
+
+  /// First row of `sobolstate` is all 1s
+  for (int64_t m = 0; m < MAXBIT; ++m) {
+    ss_a[0][m] = 1;
+  }
```
Nit
```suggestion
  std::fill=n(ss_a[0], MAXBIT, 1);
```

malfet commented on aten/src/ATen/native/SobolEngineOps.cpp:
```diff
@@ -126,19 +126,28 @@ Tensor& _sobol_engine_initialize_state_(Tensor& sobolstate, int64_t dimension) {
     int64_t m = bit_length(p) - 1;
 
+    // First m elements of row d comes from initsobolstate
     for (int64_t i = 0; i < m; ++i) {
       ss_a[d][i] = initsobolstate[d][i];
     }
```
Nit
```suggestion
    std::copy(ss_a[d], ssa_[d] + m, initsobolstate[d]);
```

Balandat replied:
I have very little understanding of the C++ internals of torch, but this fails (your suggestion above as well). Note that initsobolstate is a C array not a tensor, not sure if that's what's the issue here. I'll leave this and the `fill_n` above that also fails as is for now.
...

stderr: caffe2/aten/src/ATen/native/SobolEngineOps.cpp:141:32: error: invalid operands to binary expression ('TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' and 'int64_t' (aka 'long'))
    std::copy(ss_a[d], ss_a[d] + m, initsobolstate[d]);
                       ~~~~~~~ ^ ~
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'const at::Tensor' for 1st argument
AT_FORALL_BINARY_OPS(DEFINE_OPERATOR)
                     ^
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:102:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(const Tensor & x, Scalar y) { \
                     ^
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'const at::Tensor' for 1st argument
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:99:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(const Tensor & x, const Tensor & y) { \
                     ^
caffe2/aten/src/ATen/TensorOperators.h:110:22: note: candidate function not viable: no known conversion from 'TensorAccessor<long, 2UL - 1, DefaultPtrTraits, long>' to 'c10::Scalar' for 1st argument
caffe2/aten/src/ATen/TensorOperators.h:82:33: note: expanded from macro 'AT_FORALL_BINARY_OPS'
#define AT_FORALL_BINARY_OPS(_) \
                                ^
caffe2/aten/src/ATen/TensorOperators.h:105:22: note: expanded from macro 'DEFINE_OPERATOR'
static inline Tensor operator op(Scalar x, const Tensor & y) { \
                     ^
In module 'third_party_libgcc_std' imported from caffe2/c10/core/Allocator.h:4:
third-party-buck/platform007/build/libgcc/include/c++/7.3.0/bits/stl_algobase.h:753:29: error: cannot increment value of type 'at::TensorAccessor<long, 1, DefaultPtrTraits, long>'
           __niter > 0; --__niter, ++__first)
                                   ^ ~~~~~~~
third-party-buck/platform007/build/libgcc/include/c++/7.3.0/bits/stl_algobase.h:789:23: note: in instantiation of function template specialization 'std::__fill_n_a<at::TensorAccessor<long, 1, DefaultPtrTraits, long>, long, int>' requested here
      return _OI(std::__fill_n_a(std::__niter_base(__first), __n, __value));
                      ^
caffe2/aten/src/ATen/native/SobolEngineOps.cpp:133:8: note: in instantiation of function template specialization 'std::fill_n<at::TensorAccessor<long, 1, DefaultPtrTraits, long>, long, int>' requested here
  std::fill_n(ss_a[0], MAXBIT, 1);
       ^
2 errors generated.
    When running <c++ preprocess_and_compile>.
    When building rule //caffe2/aten:ATen-cpu#compile-pic-SobolEngineOps.cpp.o6f3daec3,platform007-clang.

malfet commented on test/test_torch.py:
```diff
@@ -1560,102 +1560,257 @@ def test_generator_cpu(self):
+        #     self.assertEqual(actual_3d.size(), torch.Size([10, 3]))
+
+        #     engine_3d = torch.quasirandom.SobolEngine(3, scramble=True, seed=1729)
+        #     draws = torch.cat([engine_3d.draw() for _ in range(0, 10)])
+        #     self.assertEqual(draws, actual_3d)
+
+        #     engine_3d = torch.quasirandom.SobolEngine(3, scramble=True, seed=1729)
+        #     engine_3d.fast_forward(5)
+        #     draws = engine_3d.draw(5)
+        #     self.assertEqual(draws, actual_3d[5:])
+        #     engine_3d.reset()
+        #     self.assertEqual(engine_3d.draw(3), actual_3d[:3])
+        #     engine_3d.fast_forward(2)
+        #     self.assertEqual(engine_3d.draw(5), actual_3d[5:])
+
+        #     # test draw_base2
+        #     engine_1d = torch.quasirandom.SobolEngine(1, scramble=True, seed=1729)
+        #     actual_1d = engine_1d.draw_base2(2)
+        #     self.assertEqual(actual_1.shape, torch.Size([4]))
+        #     engine_1d.draw_base2(2)
+        #     with self.assertRaises(ValueError):
+        #         engine_1d.draw_base2(2)
+
+        # def test_sobolengine_scrambled_lowdim_default_rng(self):
+        #     expected_1d = [0.039826, 0.484409, 0.953192, 0.799275, 0.267996]
+        #     torch.manual_seed(123456)
+        #     engine_1d = torch.quasirandom.SobolEngine(1, scramble=True)
+        #     actual_1d = engine_1d.draw(5)
+        #     self.assertEqual(actual_1d[:, 0], expected_1d)
+        #     torch.manual_seed(123456)
+        #     expected_3d = [0.133490, 0.480183, 0.855304, 0.970967, 0.345844]
+        #     engine_3d = torch.quasirandom.SobolEngine(3, scramble=True)
+        #     actual_3d = engine_3d.draw(5)
+        #     self.assertEqual(actual_3d[:, 0], expected_3d)
+
+        # def test_sobolengine_scrambled_highdim(self):
+        #     engine = torch.quasirandom.SobolEngine(1111, scramble=True)
+        #     draws = engine.draw(1000)
+        #     self.assertTrue(torch.all(draws <= 1))
+        #     self.assertTrue(torch.all(draws >= 0))
```
Please delete old test source
```suggestion

```

Balandat committed:
[SobolEngine] Update direction numbers to 21201 dims (#49710)

Summary:
Performs the update that was suggested in https://github.com/pytorch/pytorch/issues/41489

Adjust the functionality to largely match that pf the scipy companion PR https://github.com/scipy/scipy/pull/10844/, including
- a new `draw_base2` method
- include zero as the first point in the (unscrambled) Sobol sequence

The scipy PR is also quite opinionated if the `draw` method doesn't get called with a base 2 number (for which the resulting sequence has nice properties, see the scipy PR for a comprehensive discussion of this).

Note that this update is a **breaking change** in the sense that sequences generated with the same parameters after as before will not be identical! They will have the same (better, arguably) distributional properties, but calling the engine with the same seed will result in different numbers in the sequence.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/49710

Test Plan:
```
from torch.quasirandom import SobolEngine

sobol = SobolEngine(3)
sobol.draw(4)

sobol = SobolEngine(4, scramble=True)
sobol.draw(5)

sobol = SobolEngine(4, scramble=True)
sobol.draw_base2(2)
```

Reviewed By: malfet

Differential Revision: D25657233

Pulled By: Balandat

fbshipit-source-id: ce8ccafe90fc968ed811634b5f37c2eb208af985
